pkg/cgroup/v1/set.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2021 Andy Dalton
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cgroup
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/adalton/teleport-exercise/pkg/adaptation/os"
+
+	"github.com/google/uuid"
+)
+
+const (
+	DefaultBasePath                   = "/sys/fs/cgroup"
+	defaultDirectoryPerms os.FileMode = 0755
+)
+
+// Set maintains a collection of 0 or more cgroup controllers that should be
+// created/removed at the same time.
+type Set struct {
+	osAdapter   *os.Adapter
+	basePath    string
+	jobId       uuid.UUID
+	controllers []Controller
+}
+
+// NewSet creates a new cgroup (v1) set for the given jobID.  This assumes that
+// the cgroup filesystem is mounted at /sys/fs/cgroup.
+func NewSet(jobId uuid.UUID, controllers ...Controller) *Set {
+	return NewSetDetailed(nil, DefaultBasePath, jobId, controllers...)
+}
+
+// NewSetDetailed creates a new cgroup (v1) set for the given jobID rooted
+// at the given basePath.
+func NewSetDetailed(
+	osAdapter *os.Adapter,
+	basePath string,
+	jobId uuid.UUID,
+	controllers ...Controller,
+) *Set {
+
+	return &Set{
+		osAdapter:   osAdapter,
+		basePath:    basePath,
+		jobId:       jobId,
+		controllers: controllers,
+	}
+}
+
+// Create creates the cgroup v1 directories for all registered controllers.
+func (s *Set) Create() (retErr error) {
+	if s == nil {
+		// If the set is nil, then Create is vacuously successful
+		return nil
+	}
+
+	failPoint := -1
+
+	for i := range s.controllers {
+		path := fmt.Sprintf("%s/%s/jobs/%s", s.basePath, s.controllers[i].Name(), s.jobId.String())
+
+		// Create the cgroup
+		if err := s.osAdapter.MkdirAll(path, defaultDirectoryPerms); err != nil {
+			retErr = err
+			failPoint = i
+			break
+		}
+
+		// Apply the supplied configuration to the newly-created cgroup
+		if err := s.controllers[i].Apply(path); err != nil {
```

Comment:
I wonder if we could push group joining to the cgroups package, instead of having the command responsible for it. Seems clearer in terms of design.

WDYT?

Reply:
It's double, but it's open to race conditions.

The JobManager could:
* start the process, get the pid
* write the cgroup files
* send the process a signal to tell it it's safe to exec the program

I'd be possible for the child process not to have registered its signal handler before the parent sent the signal.  In that case, depending on the signal, the child might (1) terminate or (2) ignore the signal, then set up the handler, then wait forever for another signal that will never come.

It might be possible so solve that with some fancier interaction between the JobManager process and the child, but at that point the tradeoff seems not so great.

Reply:
You can move the logic to the cgroups package without changing the existing flow. This is more of a "which package should have this responsibility" question - the "when" of the setup is good.

Reply:
I could see this going either way.  The helper currently just gets a list of files and writes its pid to them, so it's not heavily coupled with cgroup implementation logic -- it's not building paths internally.

An alternative would be to pass the jobID to the program, and let it call into a cgroup API to add itself.  If we're confident that all jobs will _always_ have the same set of cgroups, then that'd work.  If we might ever want to have the flexibility to put jobs in different cgroups, then that'd make it a bit more challenging since the call wouldn't know what cgroups the calling process should be assigned to.

Reply:
Fair enough.